stan/math/rev/core/scoped_chainablestack.hpp

```diff
@@ -0,0 +1,80 @@
+#ifndef STAN_MATH_REV_CORE_SCOPED_CHAINABLESTACK_HPP
+#define STAN_MATH_REV_CORE_SCOPED_CHAINABLESTACK_HPP
+
+#include <stan/math/rev/core/chainablestack.hpp>
+
+#include <mutex>
+#include <stdexcept>
+#include <thread>
+
+namespace stan {
+namespace math {
+
+/**
+ * The AD tape of reverse mode AD is by default stored globally within the
+ * process (or thread). With the ScopedChainableStack class one may execute a
+ * nullary functor with reference to an AD tape which is stored with the
+ * instance of ScopedChainableStack. Example:
+ *
+ * ScopedChainableStack scoped_stack;
+ *
+ * double cgrad_a = scoped_stack.execute([] {
+ *   var a = 2.0;
+ *   var b = 4.0;
+ *   var c = a*a + b;
+ *   c.grad();
+ *   return a.adj();
+ * });
+ *
+ * Doing so will not interfere with the process (or thread) AD tape.
+ */
+class ScopedChainableStack {
+  ChainableStack::AutodiffStackStorage local_stack_;
+  std::mutex local_stack_mutex_;
+
+  struct activate_scope {
+    ChainableStack::AutodiffStackStorage* parent_stack_;
+    ScopedChainableStack& scoped_stack_;
+
+    explicit activate_scope(ScopedChainableStack& scoped_stack)
+        : parent_stack_(ChainableStack::instance_),
+          scoped_stack_(scoped_stack) {
+      if (!scoped_stack_.local_stack_mutex_.try_lock()) {
+        throw std::logic_error{"Cannot recurse same instance scoped stacks."};
+      }
+      ChainableStack::instance_ = &scoped_stack.local_stack_;
+    }
+
+    ~activate_scope() {
+      ChainableStack::instance_ = parent_stack_;
+      scoped_stack_.local_stack_mutex_.unlock();
+    }
+  };
+
+ public:
+  ScopedChainableStack() = default;
+
+  /**
+   * Execute in the current thread a function and write the AD
+   * tape to local_stack_ of this instance. The function may return
+   * any type.
+   *
+   * @tparam F functor to evaluate
+   * @param f instance of functor
+   * @param args arguments passed to functor
+   * @return Result of evaluated functor
+   */
+  template <typename F, typename... Args>
+  auto execute(F&& f, Args&&... args) {
```

Comment:
Usually you want `decltype(auto)` in such cases. Although in this case forwarding is not necessary at all - forwarding a functor before calling it does not do anything. So the function can just as well accept functor argument by const reference.

Reply:
ok, then let's go with `decltype(auto)` ... but then you are saying we do not need forwarding? I am not an expert on that matter (yet). When I look at the TBB doc they have these variants in a similar context:

```c++
        // Supported since C++11
        template<typename F> auto execute(F& f) -> decltype(f());
        template<typename F> auto execute(const F& f) -> decltype(f());
        template<typename F> void enqueue(F&& f);
```

What, @t4c1 , would you suggest to implement in our case?

Reply:
Sorry, my response was really not exact. For this function I would change the signature like this:
```suggestion
  auto execute(const F& f, Args&&... args) {
```
And remove the `std::forward` call. Since this is not actually ever returning references (unless f is returning a reference, but I think that is never actually needed) the return type can remain `auto`.

Reply:
Hmm... I tried the const thing, but then I can't do everything I want this thing to be able to handle. I wrote a test which fails to compile with the const thing where the functor called changes internal state of the functor. In this case the operator() is non-const and as such the `const F&` does not compile.

From comparing with how other libraries handle this a `decltype(auto)` as return type sounds reasonable to me.

... and the forward thing: I thought this is necessary to avoid calls to copy-constructors for cases when the `operator()()` functor creates internally some object which is then returned by value... no? Or does that mean we want to move here?

Reply:
I did not know modifying internal state of the functor is a realistic use case. Accepting the functor as `F&& f` is fine than.

About forwarding: Only if the functor overloads both `operator()() const&` and `operator()() &&` with different behavior. Otherwise neither `forward` nor `move` affect how this works.

Reply:
> I did not know modifying internal state of the functor is a realistic use case.

Yeah, same. What's the use case?